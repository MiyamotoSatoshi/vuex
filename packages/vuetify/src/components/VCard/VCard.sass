@import '../../styles/styles.sass'

// Theme
+theme(v-card) using ($material)
  background-color: map-get($material, 'cards')
  color: map-deep-get($material, 'text', 'primary')

  > .v-card__text
    color: map-deep-get($material, 'text', 'secondary')

  &.v-card--outlined
    border: 1px solid map-get($material, 'dividers')

// Block
.v-card
  max-width: 100%
  outline: none
  text-decoration: none
  transition-property: box-shadow, opacity
  word-break: break-word
  overflow-wrap: break-word
  +elevation(2)

  > *:first-child:not(.v-btn):not(.v-chip),
  > .v-card__progress + *:not(.v-btn):not(.v-chip)
    border-top-left-radius: inherit
    border-top-right-radius: inherit

  > *:last-child:not(.v-btn):not(.v-chip)
    border-bottom-left-radius: inherit
    border-bottom-right-radius: inherit

// Element
.v-card__progress
  top: 0
  left: 0
  right: 0
  overflow: hidden

.v-card__title
  align-items: center
  display: flex
  flex-wrap: wrap
  font-size: map-deep-get($headings, 'h5', 'size')
  font-weight: map-deep-get($headings, 'h5', 'weight')
  letter-spacing: map-deep-get($headings, 'h5', 'letter-spacing')
  line-height: map-deep-get($headings, 'h5', 'line-height')
  padding: 16px 16px 8px

  + .v-card__text
    padding-top: 0

.v-card__text
  font-size: map-deep-get($headings, 'subtitle-2', 'size')
  line-height: map-deep-get($headings, 'subtitle-2', 'line-height')
  letter-spacing: map-deep-get($headings, 'subtitle-2', 'letter-spacing')
  padding: 16px
  width: 100%

.v-card__actions
  align-items: center
  display: flex
  padding: 8px

  .v-btn.v-btn
    padding: 0 8px

    // Buttons have reduced padding in card actions per spec
    // and need to compensate the icon margins accordingly
    +ltr()
      + .v-btn
        margin-left: 8px

      .v-icon--left
        margin-left: 4px

      .v-icon--right
        margin-right: 4px

    +rtl()
      + .v-btn
        margin-right: 8px

      .v-icon--left
        margin-right: 4px

      .v-icon--right
        margin-left: 4px

// Modifier
.v-card--flat
  +elevation(0)

.v-card--hover
  cursor: pointer
  transition: box-shadow .4s cubic-bezier(.25, .8, .25, 1)

  &:hover
    +elevation(8)

.v-card--link
  cursor: pointer

  // https://github.com/vuetifyjs/vuetify/issues/3925
  .v-chip
    cursor: pointer

  &:focus:before
    opacity: 0.08

  &:before
    background: currentColor
    bottom: 0
    content: ''
    left: 0
    opacity: 0
    pointer-events: none
    position: absolute
    right: 0
    top: 0
    transition: .2s opacity map-get($transition, 'fast-in-slow-out')

.v-card--disabled
  pointer-events: none
  user-select: none

  > *:not(.v-card__progress)
    opacity: 0.6
    transition: inherit

.v-card--loading
  overflow: hidden

.v-card--outlined
  box-shadow: none

.v-card--raised
  +elevation(8)
