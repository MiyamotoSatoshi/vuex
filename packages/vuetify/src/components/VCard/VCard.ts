// Styles
import '../../stylus/components/_cards.styl'

// Extensions
import VSheet from '../VSheet'

// Mixins
import Routable from '../../mixins/routable'

// Helpers
import mixins from '../../util/mixins'

// Types
import { VNode } from 'vue'

/* @vue/component */
export default mixins(
  Routable,
  VSheet
).extend({
  name: 'v-card',

  props: {
    elevation: {
      type: [Number, String],
      default: 2
    },
    flat: Boolean,
    hover: Boolean,
    img: String,
    raised: Boolean
  },

  computed: {
    classes (): object {
      return {
        'v-card': true,
        'v-card--hover': this.hover,
        ...VSheet.options.computed.classes.call(this)
      }
    },
    computedElevation (): number | string {
      if (this.flat) return 0
      if (this.raised) return 3

      return (VSheet.options.computed as any).computedElevation.call(this)
    },
    styles (): object {
      const style = {
        ...VSheet.options.computed.styles.call(this)
      }

      if (this.img) {
        style.background = `url("${this.img}") center center / cover no-repeat`
      }

      return style
    }
  },

  render (h): VNode {
    const { tag, data } = this.generateRouteLink(this.classes)

    data.style = this.styles

    return h(tag, this.setBackgroundColor(this.color, data), this.$slots.default)
  }
})
