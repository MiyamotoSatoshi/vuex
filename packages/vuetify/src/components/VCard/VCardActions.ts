import { createSimpleFunctional } from '@/util'

export const VCardActions = createSimpleFunctional('v-card-actions')
