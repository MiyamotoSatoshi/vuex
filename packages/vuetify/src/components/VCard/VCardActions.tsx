// Composables
import { provideDefaults } from '@/composables/defaults'

// Utility
import { defineComponent } from 'vue'
import { useRender } from '@/util'

export const VCardActions = defineComponent({
  name: 'VCardActions',

  setup (_, { slots }) {
    provideDefaults({
      VBtn: {
        variant: 'text',
      },
    })

    useRender(() => (
      <div class="v-card-actions">
        { slots?.default?.() }
      </div>
    ))

    return {}
  },
})
