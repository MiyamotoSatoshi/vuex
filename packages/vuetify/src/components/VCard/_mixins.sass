@mixin card-theme($material)
  background-color: map-get($material, 'cards')
  color: map-deep-get($material, 'text', 'primary')

  .v-card__text
    color: map-deep-get($material, 'text', 'secondary')

  &.v-card--outlined
    border: 1px solid map-get($material, 'dividers')
