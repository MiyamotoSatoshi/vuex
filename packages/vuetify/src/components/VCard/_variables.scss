@import '../../styles/styles.sass';

$card-border-radius: $border-radius-root !default;
$card-shaped-border-radius: 24px !default;
$card-title-font-size: map-deep-get($headings, 'h6', 'size') !default;
$card-title-font-weight: map-deep-get($headings, 'h6', 'weight') !default;
$card-title-letter-spacing: map-deep-get($headings, 'h6', 'letter-spacing') !default;
$card-title-line-height: map-deep-get($headings, 'h6', 'line-height') !default;
$card-text-font-size: map-deep-get($headings, 'subtitle-2', 'size') !default;
$card-text-font-weight: 400 !default;
$card-text-letter-spacing: map-deep-get($headings, 'subtitle-2', 'line-height') !default;
$card-text-line-height: map-deep-get($headings, 'subtitle-2', 'letter-spacing') !default;
