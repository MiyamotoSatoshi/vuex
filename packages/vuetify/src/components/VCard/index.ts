import { createSimpleFunctional } from '../../util/helpers'
import VCard from './VCard'

const VCardActions = createSimpleFunctional('v-card__actions')
const VCardText = createSimpleFunctional('v-card__text')
const VCardTitle = createSimpleFunctional('v-card__title')

export { VCard, VCardTitle, VCardActions, VCardText }

export default {
  $_vuetify_subcomponents: {
    VCard,
    VCardTitle,
    VCardActions,
    VCardText,
  },
}
