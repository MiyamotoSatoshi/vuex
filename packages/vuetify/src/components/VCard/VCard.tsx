/* eslint-disable complexity */

// Styles
import './VCard.sass'

// Components
import { VCardActions } from './VCardActions'
import { VCardItem } from './VCardItem'
import { VCardText } from './VCardText'
import { VDefaultsProvider } from '@/components/VDefaultsProvider'
import { VImg } from '@/components/VImg'

// Directives
import { Ripple } from '@/directives/ripple'

// Composables
import { genOverlays, makeVariantProps, useVariant } from '@/composables/variant'
import { IconValue } from '@/composables/icons'
import { LoaderSlot, makeLoaderProps, useLoader } from '@/composables/loader'
import { makeBorderProps, useBorder } from '@/composables/border'
import { makeDensityProps, useDensity } from '@/composables/density'
import { makeDimensionProps, useDimension } from '@/composables/dimensions'
import { makeElevationProps, useElevation } from '@/composables/elevation'
import { makeLocationProps, useLocation } from '@/composables/location'
import { makePositionProps, usePosition } from '@/composables/position'
import { makeRoundedProps, useRounded } from '@/composables/rounded'
import { makeRouterProps, useLink } from '@/composables/router'
import { makeTagProps } from '@/composables/tag'
import { makeThemeProps, provideTheme } from '@/composables/theme'

// Utilities
import { defineComponent, useRender } from '@/util'

export const VCard = defineComponent({
  name: 'VCard',

  directives: { Ripple },

  props: {
    appendAvatar: String,
    appendIcon: IconValue,
    disabled: Boolean,
    flat: Boolean,
    hover: Boolean,
    image: String,
    link: Boolean,
    prependAvatar: String,
    prependIcon: IconValue,
    ripple: Boolean,
    subtitle: String,
    text: String,
    title: String,

    ...makeThemeProps(),
    ...makeBorderProps(),
    ...makeDensityProps(),
    ...makeDimensionProps(),
    ...makeElevationProps(),
    ...makeLoaderProps(),
    ...makeLocationProps(),
    ...makePositionProps(),
    ...makeRoundedProps(),
    ...makeRouterProps(),
    ...makeTagProps(),
    ...makeVariantProps({ variant: 'elevated' } as const),
  },

  setup (props, { attrs, slots }) {
    const { themeClasses } = provideTheme(props)
    const { borderClasses } = useBorder(props)
    const { colorClasses, colorStyles, variantClasses } = useVariant(props)
    const { densityClasses } = useDensity(props)
    const { dimensionStyles } = useDimension(props)
    const { elevationClasses } = useElevation(props)
    const { loaderClasses } = useLoader(props)
    const { locationStyles } = useLocation(props)
    const { positionClasses } = usePosition(props)
    const { roundedClasses } = useRounded(props)
    const link = useLink(props, attrs)

    useRender(() => {
      const Tag = (link.isLink.value) ? 'a' : props.tag
      const hasTitle = !!(slots.title || props.title)
      const hasSubtitle = !!(slots.subtitle || props.subtitle)
      const hasHeader = hasTitle || hasSubtitle
      const hasAppend = !!(slots.append || props.appendAvatar || props.appendIcon)
      const hasPrepend = !!(slots.prepend || props.prependAvatar || props.prependIcon)
      const hasImage = !!(slots.image || props.image)
      const hasCardItem = hasHeader || hasPrepend || hasAppend
      const hasText = !!(slots.text || props.text)
      const isClickable = !props.disabled && (link.isClickable.value || props.link)

      return (
        <Tag
          class={[
            'v-card',
            {
              'v-card--disabled': props.disabled,
              'v-card--flat': props.flat,
              'v-card--hover': props.hover && !(props.disabled || props.flat),
              'v-card--link': isClickable,
            },
            themeClasses.value,
            borderClasses.value,
            colorClasses.value,
            densityClasses.value,
            elevationClasses.value,
            loaderClasses.value,
            positionClasses.value,
            roundedClasses.value,
            variantClasses.value,
          ]}
          style={[
            colorStyles.value,
            dimensionStyles.value,
            locationStyles.value,
          ]}
          href={ link.href.value }
          onClick={ isClickable && link.navigate }
          v-ripple={ isClickable }
        >
          { hasImage && (
            <VDefaultsProvider
              key="image"
              defaults={{
                VImg: {
                  cover: true,
                  src: props.image,
                },
              }}
            >
              <div class="v-card__image">
                { slots.image?.() ?? <VImg /> }
              </div>
            </VDefaultsProvider>
          ) }

          <LoaderSlot
            name="v-card"
            active={ !!props.loading }
            color={ typeof props.loading === 'boolean' ? undefined : props.loading }
            v-slots={{ default: slots.loader }}
          />

          { hasCardItem && (
            <VCardItem
              key="item"
              prependAvatar={ props.prependAvatar }
              prependIcon={ props.prependIcon }
              title={ props.title }
              subtitle={ props.subtitle }
              appendAvatar={ props.appendAvatar }
              appendIcon={ props.appendIcon }
            >
              {{
                default: slots.item,
                prepend: slots.prepend,
                title: slots.title,
                subtitle: slots.subtitle,
                append: slots.append,
              }}
            </VCardItem>
          ) }

          { hasText && (
            <VCardText key="text">
              { slots.text?.() ?? props.text }
            </VCardText>
          ) }

          { slots.default?.() }

          { slots.actions && (
            <VCardActions v-slots={{ default: slots.actions }} />
          ) }

          { genOverlays(isClickable, 'v-card') }
        </Tag>
      )
    })

    return {}
  },
})

export type VCard = InstanceType<typeof VCard>
