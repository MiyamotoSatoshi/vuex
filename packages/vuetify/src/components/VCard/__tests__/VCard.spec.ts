import {
  mount,
  Wrapper,
  MountOptions,
} from '@vue/test-utils'
import VCard from '../VCard'
import { ExtractVue } from '../../../util/mixins'

describe('VCard.vue', () => {
  type Instance = ExtractVue<typeof VCard>
  let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>
  beforeEach(() => {
    mountFunction = (options?: MountOptions<Instance>) => {
      return mount(VCard, options)
    }
  })

  it('should render component and match snapshot', () => {
    const wrapper = mountFunction()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render loading card', () => {
    const wrapper = mountFunction({
      propsData: {
        loading: true,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render card, which is link', () => {
    const wrapper = mountFunction({
      listeners: {
        click: () => {},
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render card with img', () => {
    const wrapper = mountFunction({
      propsData: {
        img: 'image.jpg',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render a flat card', () => {
    const wrapper = mountFunction({
      propsData: {
        flat: true,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render a raised card', () => {
    const wrapper = mountFunction({
      propsData: {
        raised: true,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render a card with custom height', () => {
    const heightpx = '400px'
    const wrapper = mountFunction({
      propsData: {
        height: heightpx,
      },
    })

    expect(wrapper.element.style.height).toBe(heightpx)
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setProps({
      height: 401,
    })
    expect(wrapper.element.style.height).toBe('401px')
  })
})
