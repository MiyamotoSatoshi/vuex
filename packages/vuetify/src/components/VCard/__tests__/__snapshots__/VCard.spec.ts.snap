// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VCard should match a snapshot 1`] = `
<div class="v-card v-theme--light v-card--density-default v-card--variant-contained">
</div>
`;

exports[`VCard should render slots and match a snapshot 1`] = `
<div class="v-card v-theme--light v-card--density-default v-card--variant-contained">
  <div class="v-card-img">
    <v-img src="foo.png">
    </v-img>
  </div>
  <div class="v-card-media">
    <div>
      bar
    </div>
  </div>
  <div class="v-card-header">
    <div class="v-card-avatar">
      <div>
        foobar
      </div>
    </div>
    <div class="v-card-header-text">
      <div class="v-card-title">
        <div>
          fizzbuzz
        </div>
      </div>
      <div class="v-card-subtitle">
        <div>
          fizz
        </div>
      </div>
    </div>
    <div class="v-card-avatar">
      <div>
        foo
      </div>
    </div>
  </div>
  <div class="v-card-text">
    <div>
      buzz
    </div>
  </div>
  <div class="v-card-actions">
    <div>
      actions
    </div>
  </div>
</div>
`;
